File: golang/test/hmac_test.go
package test

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/require"
)

const (
	// TODO: HMAC salt - In actual implementation, load from environment variables or configuration file
	salt = "my_secret_salt_value_!@#$%^&*" // hmac key
)

type Body struct {
	UserID    int    `json:"userId"`
	Username  string `json:"username"`
	Email     string `json:"email"`
	Role      string `json:"role"`
	CreatedAt int    `json:"createdAt"`
}

var (
	body = Body{
		UserID:    1234,
		Username:  "홍길동",
		Email:     "user@example.com",
		Role:      "admin",
		CreatedAt: 1234567890,
	}
)

func TestSha256(t *testing.T) {
	bodyBytes, err := json.Marshal(body)
	require.NoError(t, err)
	t.Log(string(bodyBytes))

	hmac := hmac.New(sha256.New, []byte(salt))
	hmac.Write(bodyBytes)
	hashBytes := hmac.Sum(nil)
	hashString := hex.EncodeToString(hashBytes)
	t.Log("hashString", hashString) // expected X-HMAC-Signature: f96cf60394f6b8ad3c6de2d5b2b1d1a540f9529082a8eb9cee405bfbdd9f37a1
}

File: typescript/test/hmac.test.ts
import { HmacService } from '../src/services/hmacService';

// TODO: HMAC salt - In actual implementation, load from environment variables or configuration file
const salt = "my_secret_salt_value_!@#$%^&*"; // hmac key

interface Body {
  userId: number;
  username: string;
  email: string;
  role: string;
  createdAt: number;
}

const body: Body = {
  userId: 1234,
  username: "홍길동",
  email: "user@example.com",
  role: "admin",
  createdAt: 1234567890,
};

describe('HMAC Service', () => {
  test('should generate correct HMAC signature', () => {
    const bodyBytes = JSON.stringify(body);
    console.log(bodyBytes);
    
    const hashString = HmacService.generateHmac(body);
    console.log('hashString:', hashString); // expected X-HMAC-Signature: f96cf60394f6b8ad3c6de2d5b2b1d1a540f9529082a8eb9cee405bfbdd9f37a1
    
    expect(hashString).toBeDefined();
    expect(hashString.length).toBe(64); // SHA256 hex string length
  });

  test('should validate HMAC signature correctly', () => {
    const bodyString = JSON.stringify(body);
    const signature = HmacService.generateHmacFromString(bodyString);
    
    expect(HmacService.validateHmac(bodyString, signature)).toBe(true);
    expect(HmacService.validateHmac(bodyString, 'invalid_signature')).toBe(false);
  });
}); 